File: src/addConnectionDialog.js
import GObject from "gi://GObject";
import Gtk from "gi://Gtk";
import Adw from "gi://Adw";
import Tlg from 'gi://Tlg';

export class AddConnectionDialog extends Adw.Window {
    constructor(mainWindow) {
        super({ transient_for: mainWindow, modal: true });

        this.validateInput();
    }

    static {
        GObject.registerClass({
            GTypeName: 'TrilogyAddConnectionDialog',
            Properties: {
                "valid": GObject.ParamSpec.boolean(
                    "valid",
                    "Valid",
                    "Whether the input is valid",
                    GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT,
                    false
                )
            },
            Signals: {
                'add-connection': {
                    param_types: [Tlg.Connection]
                }
            },
            Template:  "resource:///com/rutins/Trilogy/addConnectionDialog.ui",
            InternalChildren: ["nameEntry", "urlEntry"]
        }, this);
    }

    get valid() {
        return this._valid ?? false;
    }

    set valid(value) {
        if(this._valid == value)
            return;

        this._valid = value;
        this.notify('valid');
    }

    cancelClicked() {
        this.close();
    }

    addClicked() {
        this.emit('add-connection', new Tlg.Connection({
            name: this._nameEntry.get_text(),
            url: this._urlEntry.get_text()
        }));
        this.close();
    }

    validateInput() {
        const name = this._nameEntry.get_text();
        const url = this._urlEntry.get_text();
        let failed = false;
        if(name.includes(' ') || !name) {
            failed = true;
            this._nameEntry.add_css_class("error");
        } else {
            this._nameEntry.remove_css_class("error");
        }

        if(!url.startsWith('postgresql://') || !url) {
            failed = true;
            this._urlEntry.add_css_class("error");
        } else {
            this._urlEntry.remove_css_class("error");
        }

        this.valid = !failed;
    }
}

File: src/add_connection_dialog.py
from gi.repository import Gtk, Adw, GObject, Tlg

@Gtk.Template(resource_path='/com/rutins/Trilogy/add-connection-dialog.ui')
class AddConnectionDialog(Adw.Window):
    __gtype_name__ = 'TrilogyAddConnectionDialog'
    __gsignals__ = {
        'add_connection': (GObject.SIGNAL_RUN_LAST, GObject.TYPE_NONE, (Tlg.Connection,))
    }

    name_entry = Gtk.Template.Child()
    url_entry = Gtk.Template.Child()

    valid = GObject.Property(type=bool, default=False)

    def __init__(self, main_window):
        super().__init__(transient_for=main_window, modal=True)
        self.validate_input()
        
    @Gtk.Template.Callback()
    def cancel_clicked(self, *args):
        self.close()
    
    @Gtk.Template.Callback()
    def add_clicked(self, *args):
        self.emit('add-connection', Tlg.Connection(
            name=self.name_entry.get_text(),
            url=self.url_entry.get_text()
        ))
        self.close()
    
    @Gtk.Template.Callback()
    def validate_input(self, *args):
        name: str = self.name_entry.get_text()
        url: str = self.url_entry.get_text()
        failed = False

        if name == '':
            failed = True
            self.name_entry.add_css_class('error')
        else:
            self.name_entry.remove_css_class('error')

        if Tlg.protocol_for_uri(url) == Tlg.Protocol.UNKNOWN:
            failed = True
            self.url_entry.add_css_class('error')
        else:
            self.url_entry.remove_css_class('error')
        
        self.props.valid = not failed
